File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    // Optimal Approach : O(n)
    public int subarraySum(int[] nums, int k) {
       int n = nums.length;
       // we will use prefix sum here {preSum[j]-preSum[i] = k} => {preSum[i] = preSum[j]-k}

       HashMap<Integer,Integer> mp = new HashMap<>();

       // put {0,1}==> if sum = 0 then freq = 1;
       mp.put(0,1);
        int preSum = 0;
        int res = 0;
       for(int val : nums){
            preSum += val;

            if(mp.containsKey(preSum - k)){
                res += mp.get(preSum - k);
            }
            mp.put(preSum, mp.getOrDefault(preSum, 0)+1);
       }
       return res;

    }
}

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
         int n = nums.size();
        unordered_map<int,int> mp;
        mp[0] = 1;
        int ans = 0;
        int currSm = 0;
        for(int i = 0; i<n; i++){
            currSm += nums[i];
            if(mp.find(currSm - k) != mp.end()){
                ans += mp[currSm-k];
            }
            mp[currSm]++;

        }
        return ans;
    }
};